### PR TITLE
fabtests: add device support infrastructure

### PR DESCRIPTION
**Pull request by aingerson**
Initial device support for fabtests with L0 support

Changes:
- use fi_mr_regattr to register device memory
- general hmem_ops infrastructure mimicking libfabric structure
- add L0 configure check
- add L0 hmem ops for ZE testing

### CODE REVIEW

**Comment by aingerson** [comment]
@shefty @iziemba @rajachan 
Same thing here. Still a work in progress. Can split out.

**Comment by iziemba** [comment]
@aingerson Sorry I accidentally closed the pull request! Clicked the wrong button....

Would it make sense to break out the core changes into a separate PR from Intel GPU specific implementation?

**Comment by aingerson** [comment]
@iziemba No worries, definitely been there before...
Yeah sure I can do that if you think it would be easier. Mostly just put it in one place for initial review and since the L0 changes depend on the general ones but I can split it up so we can get the general ones merged for other development. @shefty preference?

File: fabtests/common/shared.c [hunk 1]
```diff
@@ -380,20 +406,19 @@ static int ft_alloc_ctx_array(struct ft_context **mr_array, char ***mr_bufs,
 		context = &(*mr_array)[i];
 		if (!(opts.options & FT_OPT_ALLOC_MULT_MR)) {
 			context->buf = default_buf + mr_size * i;
+			context->mr = mr;
+			context->desc = mr_desc;
```
Comment (shefty) [inline fabtests/common/shared.c:427]: Why this change?

File: fabtests/common/shared.c [hunk 2]
```diff
@@ -402,6 +405,12 @@ static int ft_alloc_ctx_array(struct ft_context **mr_array, char ***mr_bufs,
 			return -FI_ENOMEM;
 	}
 
+	reg = ((fi->domain_attr->mr_mode & FI_MR_LOCAL &&
+		!(opts.options & FT_OPT_USE_DEVICE)) ||
+		(fi->domain_attr->mr_mode & FI_MR_HMEM &&
+		opts.options & FT_OPT_USE_DEVICE) ||
+		fi->caps & (FI_RMA | FI_ATOMIC));
```
Comment (shefty) [inline fabtests/common/shared.c:420]: Please pull this check into it's own function.

File: fabtests/common/shared.c [hunk 3]
```diff
@@ -494,7 +507,10 @@ static int ft_alloc_msgs(void)
 	remote_cq_data = ft_init_cq_data(fi);
 
 	if (!ft_mr_alloc_func && !ft_check_opts(FT_OPT_SKIP_REG_MR) &&
-	    (fi->domain_attr->mr_mode & FI_MR_LOCAL ||
+	    ((fi->domain_attr->mr_mode & FI_MR_LOCAL &&
+	      !(opts.options & FT_OPT_USE_DEVICE)) ||
+	    (fi->domain_attr->mr_mode & FI_MR_HMEM &&
+	     opts.options & FT_OPT_USE_DEVICE) ||
```
Comment (shefty) [inline fabtests/common/shared.c:521]: Call new function here.

File: fabtests/common/shared.c [hunk 4]
```diff
@@ -662,12 +678,21 @@ int ft_alloc_active_res(struct fi_info *fi)
...
 {
+	int ret;
+
 	tx_seq = 0;
 	rx_seq = 0;
 	tx_cq_cntr = 0;
 	rx_cq_cntr = 0;
+
+	if (opts.options & FT_OPT_USE_DEVICE) {
+		ret = ft_hmem_init(opts.iface);
```
Comment (shefty) [inline fabtests/common/shared.c:699]: Should always call hmem_init, since we always call hmem_alloc above.

File: fabtests/common/shared.c [hunk 5]
```diff
@@ -1147,9 +1175,16 @@ int ft_join_mc(void)
...
 {
+	char tmp_buf[FT_MAX_CTRL_MSG];
 	int ret;
 
-	ret = fi_av_insert(av, addr, count, fi_addr, flags, context);
+	ret = ft_hmem_copy_from(opts.iface, opts.device, tmp_buf, addr, FT_MAX_CTRL_MSG);
+	if (ret) {
+		FT_PRINTERR("ft_hmem_copy_from", ret);
+		return ret;
+	}
```
Comment (shefty) [inline fabtests/common/shared.c:1193]: It's unclear why this change is needed (yes, I get it, but a future reader probably won't).  Maybe we should force OOB address exchange if hmem is used.

**Comment by shefty** [comment]
If you can split out the core changes in this case, it would help the parallel development effort that's going on.  Under normal situations, I'm fine with them being part of the same PR.

**Comment by aingerson** [comment]
@shefty @iziemba Updated to remove L0 changes and address comments. I'll open a separate PR for the L0 changes and use the super fancy new label to depend on this PR.

**Comment by aingerson** [comment]
> We really need a way to share code between libfabric and fabtests. That should be easy with them in the same repo.

I agree. Then we can share both the hmem ops as well as the atomics. I'll look into this and propose something.

File: fabtests/common/shared.c [hunk 6]
```diff
@@ -664,12 +674,20 @@ int ft_alloc_active_res(struct fi_info *fi)
...
 	tx_seq = 0;
 	rx_seq = 0;
 	tx_cq_cntr = 0;
 	rx_cq_cntr = 0;
+
+	//If using device memory for transfers, require OOB address
+	//exchange because extra steps are involved when passing
+	//device buffers into fi_av_insert
+	if (opts.options & FT_OPT_USE_DEVICE)
+		opts.options |= FT_OPT_OOB_ADDR_EXCH;
```
Comment (iziemba) [inline fabtests/common/shared.c:688]: Not quite sure I follow. Is this to avoid passing a device buffer into fi_av_insert?

Comment (shefty) [reply fabtests/common/shared.c:688]: Yes, this change was made in response to my comments.  The previous patch had code to copy the address data that was received into a message into a host bounce buffer for AV handling.  For someone looking at the code, the need to use a bounce buffer isn't obvious.  This change removes the bounce buffer and reverts the AV code back to the current code.

**Comment by shefty** [comment]
Can I get a quick status on this PR?

**Comment by aingerson** [comment]
@shefty Great question... not sure what the OSX or windows errors are... Mac error just looks like some Travis issue. Windows error is probably some other file I forgot to update but I'm not sure which one. Any idea?
I can investigate those, but otherwise this PR should be good to go.

**Comment by aingerson** [comment]
@shefty reran CIs and looks like Mac issue was a Travis issue/timeout. Windows error still have no clue. Do you?

**Comment by shefty** [comment]
@aingerson - check Makefile.win.  You probably need to add hmem.c to whichever tests now support hmem.

**Comment by aingerson** [comment]
@shefty ahhh thank you! Updated
